File: sdk/js/test/authority.test.ts
// Authority binding in TypeScript [TAP-EVT-AUTHORIZATION, §9.2, PROVISIONAL],
// mirroring sdk/python/tests/test_authority.py.
//
// Before this, `authorization`/`authority_state_version`/`target_state_digest`
// appeared nowhere in the TypeScript tree — only tap_ref.py and the Python SDK
// implemented the stateless half of §9.2, so a TS Signer could not bind an
// Event to a declared approval, and a TS Verifier could not label the outcome.
//
//   npx tsx test/authority.test.ts

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

import {
  Authorization,
  AuthorityExpired,
  AuthorityRevoked,
  authorityEffectLabel,
  authorityStateVersion,
  checkAuthorityWindow,
  checkAuthorityNotRevoked,
} from "../src/authority.js";
import { jsonDigest, newId, publicJwk, signEvent, signPassport, textDigest } from "../src/tap.js";
import { TAPClient } from "../src/client.js";
import { evaluateEvent, verifyTranscript } from "../src/verify.js";

const AUTHORITY_STATE = { policy: "refund-v3", rules: ["max_refund_cents:10000"] };
const TARGET_STATE = { ticket: "8842", status: "refunded", refund_cents: 5000 };

const here = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(readFileSync(resolve(here, "../../../test-vectors.json"), "utf8"));

let n = 0;
const tests: [string, () => Promise<void>][] = [];
const test = (name: string, fn: () => Promise<void>) => tests.push([name, fn]);
const assert = (cond: unknown, msg: string) => {
  if (!cond) throw new Error(msg);
};

test("authorityStateVersion is the same JCS-digest construction as jsonDigest", async () => {
  // Same construction as the Python SDK's authority_state_version / policy_version:
  // sha256(JCS(state)) — jsonDigest already performs exactly that.
  assert(
    authorityStateVersion(AUTHORITY_STATE) === jsonDigest(AUTHORITY_STATE),
    "authorityStateVersion must equal jsonDigest for the same input",
  );
});

test("grant() and toRecord()/fromRecord() round-trip", async () => {
  const auth = Authorization.grant({
    targetState: TARGET_STATE,
    authorityState: AUTHORITY_STATE,
    nbf: 1000,
    exp: 2000,
    issuerKid: "key_admin",
  });
  const record = auth.toRecord();
  assert(
    record.authority_state_version === authorityStateVersion(AUTHORITY_STATE),
    "authority_state_version must match",
  );
  assert(record.target_state_digest === jsonDigest(TARGET_STATE), "target_state_digest must match");
  assert(record.nbf === 1000 && record.exp === 2000, "nbf/exp must round-trip");
  assert(record.issuer_kid === "key_admin", "issuer_kid must round-trip");
  const back = Authorization.fromRecord(record);
  assert(JSON.stringify(back.toRecord()) === JSON.stringify(record), "fromRecord must round-trip exactly");
});

test("checkAuthorityWindow enforces the +/-60s skew boundary", async () => {
  const auth = Authorization.grant({
    targetState: TARGET_STATE,
    authorityState: AUTHORITY_STATE,
    nbf: 1000,
    exp: 2000,
    issuerKid: "key_admin",
  });
  checkAuthorityWindow(auth, 1500); // mid-window: must not throw
  checkAuthorityWindow(auth, 1000 - 60); // inclusive lower skew edge: must not throw
  for (const badNow of [1000 - 61, 2000 + 61]) {
    let threw = false;
    try {
      checkAuthorityWindow(auth, badNow);
    } catch (e) {
      threw = e instanceof AuthorityExpired;
      if (threw) {
        assert((e as AuthorityExpired).authorization.authz_id === auth.authzId, "authz_id carried on the error");
        assert((e as AuthorityExpired).now === badNow, "now carried on the error");
      }
    }
    assert(threw, `expected AuthorityExpired at now=${badNow}`);
  }
});

test("authorityEffectLabel matches/mismatches/unverifies correctly", async () => {
  const auth = Authorization.grant({
    targetState: TARGET_STATE,
    authorityState: AUTHORITY_STATE,
    nbf: 1000,
    exp: 2000,
    issuerKid: "key_admin",
  });
  assert(authorityEffectLabel(null, { effect_digest: "x" }) === null, "no authorization -> null");
  assert(authorityEffectLabel(auth, {}) === "unverified_authority", "no effect_digest -> unverified");
  assert(
    authorityEffectLabel(auth, { effect_digest: jsonDigest(TARGET_STATE) }) === "authorized_match",
    "matching effect_digest -> authorized_match",
  );
  const other = { ...TARGET_STATE, refund_cents: 7500 };
  assert(
    authorityEffectLabel(auth, { effect_digest: jsonDigest(other) }) === "authorized_mismatch",
    "differing effect_digest -> authorized_mismatch",
  );
  // record form (as it appears on a decoded wire Event) works identically to the class instance
  assert(
    authorityEffectLabel(auth.toRecord(), { effect_digest: jsonDigest(TARGET_STATE) }) === "authorized_match",
    "record form behaves identically to the class instance",
  );
});

test("reproduces test-vectors.json -> authorization's fixed validity_window and effect_check", async () => {
  // The vector's own _about note: "fixture data for the validity-window
  // boundaries and the match/mismatch/unverified effect labels, so other
  // implementations have fixed numbers to test against without needing their
  // own seed." Cross-checking against these SAME numbers (not just our own
  // freely-chosen nbf/exp) is what actually proves TS and Python agree on
  // [TAP-AUTHORITY-VALIDITY] and [TAP-AUTHORITY-EFFECT], not merely that each
  // independently satisfies its own logic.
  const av = vectors.authorization;
  const auth = Authorization.fromRecord(av.signed_event.authorization);
  const vw = av.validity_window;

  checkAuthorityWindow(auth, vw.accepted_now); // must not throw

  for (const badNow of [vw.rejected_now_not_yet_valid, vw.rejected_now_expired]) {
    let threw = false;
    try {
      checkAuthorityWindow(auth, badNow);
    } catch (e) {
      threw = e instanceof AuthorityExpired;
    }
    assert(threw, `vector's rejected_now=${badNow} must raise AuthorityExpired`);
  }

  const ec = av.effect_check;
  assert(
    authorityEffectLabel(auth, { effect_digest: ec.authorized_match_effect_digest }) === "authorized_match",
    "the vector's authorized_match_effect_digest must label authorized_match",
  );
  assert(
    authorityEffectLabel(auth, { effect_digest: ec.authorized_mismatch_effect_digest }) === "authorized_mismatch",
    "the vector's authorized_mismatch_effect_digest must label authorized_mismatch",
  );
  assert(
    authorityEffectLabel(auth, {}) === "unverified_authority",
    "an absent effect_digest must label unverified_authority",
  );
});

test("traceTool binds authorization and stamps effect_digest", async () => {
  const captured: Record<string, unknown>[] = [];
  const tap = new TAPClient({
    agentId: "bot",
    privateKeyHex: "44".repeat(32),
    kid: "key_test",
    post: async (_url, payload: any) => {
      captured.push(...(payload.events ?? []));
    },
  });
  await tap.issuePassport({ taskPrompt: "t", scope: ["call:tool"] });
  const auth = tap.authorize({ targetState: TARGET_STATE, authorityState: AUTHORITY_STATE, ttlS: 3600 });

  await tap.traceTool(
    { tool: "issue_refund", scope: "call:tool", authorize: auth, effectOf: () => TARGET_STATE },
    async () => TARGET_STATE,
  );
  await tap.flush();

  const ev = captured[0] as any;
  assert(ev.authorization.authz_id === auth.authzId, "event carries the bound authz_id");
  assert(ev.result.effect_digest === jsonDigest(TARGET_STATE), "event carries the stamped effect_digest");
  assert(
    authorityEffectLabel(ev.authorization, ev.result) === "authorized_match",
    "the stamped Event labels as authorized_match",
  );
});

test("traceTool blocks before acting on an expired authorization", async () => {
  // Same inline-enforcement shape as a policy denial: the wrapped fn never
  // runs, and a signed `denied` event with DENIED_AUTHORITY_EXPIRED is still emitted.
  const captured: Record<string, unknown>[] = [];
  const tap = new TAPClient({
    agentId: "bot",
    privateKeyHex: "44".repeat(32),
    kid: "key_test",
    post: async (_url, payload: any) => {
      captured.push(...(payload.events ?? []));
    },
  });
  await tap.issuePassport({ taskPrompt: "t", scope: ["call:tool"] });
  const now = Math.floor(Date.now() / 1000);
  const expired = tap.authorize({
    targetState: TARGET_STATE,
    authorityState: AUTHORITY_STATE,
    ttlS: 1,
    issuerKid: "key_test",
  });
  // force well outside the skew window (Authorization fields are readonly by
  // design — reconstruct rather than mutate, matching the intent of the guard)
  const forcedExpired = Authorization.fromRecord({ ...expired.toRecord(), exp: now - 3600 });

  let ran = false;
  let threw = false;
  try {
    await tap.traceTool(
      { tool: "issue_refund", scope: "call:tool", authorize: forcedExpired },
      async () => {
        ran = true;
        return TARGET_STATE;
      },
    );
  } catch (e) {
    threw = e instanceof AuthorityExpired;
  }

  await tap.flush();
  assert(threw, "expected AuthorityExpired to propagate");
  assert(!ran, "the wrapped function must not run");
  assert((captured[0] as any).action.kind === "denied", "a denied event is emitted");
  assert((captured[0] as any).result.code === "DENIED_AUTHORITY_EXPIRED", "denied with the correct code");
  assert(
    (captured[0] as any).authorization.authz_id === forcedExpired.authzId,
    "the denied event still carries the authz_id",
  );
});

test("checkAuthorityNotRevoked enforces the effective-time boundary", async () => {
  checkAuthorityNotRevoked("auz_x", "sha256:v", null, 1500); // nothing revoked: must not throw
  checkAuthorityNotRevoked("auz_x", "sha256:v", 2000, 1999); // before boundary: must not throw
  for (const badTs of [2000, 2001]) {
    let threw = false;
    try {
      checkAuthorityNotRevoked("auz_x", "sha256:v", 2000, badTs);
    } catch (e) {
      threw = e instanceof AuthorityRevoked;
      if (threw) {
        const err = e as AuthorityRevoked;
        assert(err.revokedId === "auz_x" && err.revokedAt === 2000 && err.signedTs === badTs, "error fields carried");
      }
    }
    assert(threw, `expected AuthorityRevoked at signedTs=${badTs}`);
  }
});

const REVOKE_SEED = "88".repeat(32);
const REVOKE_KID = "key_revoke_ts_test";

async function signedAuthorizedEvent(args: { authzId: string; eventId: string; seq: number; actionRef: string }) {
  const jwk = await publicJwk(REVOKE_SEED, REVOKE_KID);
  const passportClaims = {
    iss: "https://verifier.example.com", iat: 1000, exp: 2000,
    jti: newId("psp"), aid: "agt_revoke_ts_test",
    cid: "sha256:" + "0".repeat(64), scope: ["call:tool"], meta: {},
  };
  const passportJwt = await signPassport(REVOKE_SEED, REVOKE_KID, passportClaims);
  const auth = Authorization.grant({
    targetState: TARGET_STATE, authorityState: AUTHORITY_STATE,
    nbf: 1000, exp: 999999999, issuerKid: REVOKE_KID, authzId: args.authzId,
  });
  const body = {
    v: "tap/0.1", event_id: args.eventId, passport_jti: passportClaims.jti,
    aid: passportClaims.aid, cid: passportClaims.cid, seq: args.seq,
    ts: new Date().toISOString(),
    action: { kind: "tool_call", intent_digest: textDigest("t"), tool: "issue_refund", scope_used: "call:tool" },
    authorization: auth.toRecord(),
    result: { status: "success", code: "OK", error: null },
    attestor: "agent", action_ref: args.actionRef, kid: REVOKE_KID,
  };
  const event = await signEvent(REVOKE_SEED, body);
  return { event, passportJwt, jwk };
}

test("evaluateEvent surfaces revocation via a supplied resolver", async () => {
  const { event, jwk } = await signedAuthorizedEvent(
    { authzId: "auz_revoke_1", eventId: "evt_revoke_1", seq: 1, actionRef: "act_revoke_1" });

  // No resolver supplied: the revocation signal stays false, not an error.
  let res = await evaluateEvent(event, { resolveKey: async () => jwk, passportClaims: null, lastSeq: null });
  assert(res.sig_valid && res.authority_revoked === false, "no resolver -> not revoked");

  // A resolver that knows about this authz_id, revoked in the past: flagged.
  res = await evaluateEvent(event, {
    resolveKey: async () => jwk, passportClaims: null, lastSeq: null,
    resolveAuthorityRevocation: async (authzId) => (authzId === "auz_revoke_1" ? 1 : null),
  });
  assert(res.authority_revoked === true, "resolver flags revocation");
  assert(res.sig_valid === true, "revocation must not flip sig_valid");

  // A resolver that has never heard of this id: not flagged.
  res = await evaluateEvent(event, {
    resolveKey: async () => jwk, passportClaims: null, lastSeq: null,
    resolveAuthorityRevocation: async () => null,
  });
  assert(res.authority_revoked === false, "unknown to the resolver -> not revoked");
});

test("verifyTranscript detects batch-local authz_id reuse", async () => {
  const { event: ev1, passportJwt, jwk } = await signedAuthorizedEvent(
    { authzId: "auz_reused", eventId: "evt_reuse_1", seq: 1, actionRef: "act_reuse_1" });
  const { event: ev2 } = await signedAuthorizedEvent(
    { authzId: "auz_reused", eventId: "evt_reuse_2", seq: 2, actionRef: "act_reuse_2" });

  let report = await verifyTranscript(passportJwt, [ev1, ev2], { resolveKey: async () => jwk });
  assert(report.authority_reuse.length === 1, "one reuse detected");
  const reuse = report.authority_reuse[0] as any;
  assert(reuse.authz_id === "auz_reused");
  assert(
    new Set([reuse.first_event_id, reuse.reused_event_id]).size === 2 &&
      [reuse.first_event_id, reuse.reused_event_id].includes("evt_reuse_1") &&
      [reuse.first_event_id, reuse.reused_event_id].includes("evt_reuse_2"),
    "both event ids named",
  );

  report = await verifyTranscript(passportJwt, [ev1], { resolveKey: async () => jwk });
  assert(report.authority_reuse.length === 0, "a single event carrying the id is not reuse");
});

async function main() {
  console.log(`1..${tests.length}`);
  for (const [name, fn] of tests) {
    try {
      await fn();
    } catch (e) {
      console.log(`not ok ${++n} - ${name}`);
      console.log(`  # ${(e as Error).message}`);
      process.exit(1);
    }
    console.log(`ok ${++n} - ${name}`);
  }
  console.log("# authority binding matches tap_ref.py / Python SDK (provisional, §9.2) ✓");
}

main();

File: sdk/go/tap/authority_test.go
package tap

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// Authority binding in Go [TAP-EVT-AUTHORIZATION, §9.2, PROVISIONAL], mirroring
// sdk/python/tests/test_authority.py and sdk/js/test/authority.test.ts.
//
// Before this, authorization/authority_state_version/target_state_digest
// appeared nowhere in the Go tree — only tap_ref.py, the Python SDK, and the
// TypeScript SDK implemented the stateless half of §9.2.

var authorityStateFixture = map[string]any{
	"policy": "refund-v3",
	"rules":  []any{"max_refund_cents:10000"},
}

var targetStateFixture = map[string]any{
	"ticket": "8842", "status": "refunded", "refund_cents": float64(5000),
}

func TestAuthorityStateVersionMatchesJSONDigest(t *testing.T) {
	got, err := AuthorityStateVersion(authorityStateFixture)
	if err != nil {
		t.Fatalf("AuthorityStateVersion: %v", err)
	}
	want, err := JSONDigest(authorityStateFixture)
	if err != nil {
		t.Fatalf("JSONDigest: %v", err)
	}
	if got != want {
		t.Errorf("AuthorityStateVersion = %q, want %q", got, want)
	}
}

func TestGrantAndRecordRoundTrip(t *testing.T) {
	auth, err := GrantAuthorization(targetStateFixture, authorityStateFixture, 1000, 2000, "key_admin", "")
	if err != nil {
		t.Fatalf("GrantAuthorization: %v", err)
	}
	if auth.Nbf != 1000 || auth.Exp != 2000 || auth.IssuerKID != "key_admin" {
		t.Errorf("unexpected fields: %+v", auth)
	}
	wantASV, _ := AuthorityStateVersion(authorityStateFixture)
	if auth.AuthorityStateVersion != wantASV {
		t.Errorf("authority_state_version = %q, want %q", auth.AuthorityStateVersion, wantASV)
	}
	wantTSD, _ := JSONDigest(targetStateFixture)
	if auth.TargetStateDigest != wantTSD {
		t.Errorf("target_state_digest = %q, want %q", auth.TargetStateDigest, wantTSD)
	}

	back, err := AuthorizationFromRecord(auth.ToRecord())
	if err != nil {
		t.Fatalf("AuthorizationFromRecord: %v", err)
	}
	if *back != *auth {
		t.Errorf("round trip mismatch: %+v != %+v", *back, *auth)
	}
}

func TestCheckAuthorityWindowSkewBoundary(t *testing.T) {
	auth, err := GrantAuthorization(targetStateFixture, authorityStateFixture, 1000, 2000, "key_admin", "")
	if err != nil {
		t.Fatalf("GrantAuthorization: %v", err)
	}
	if err := CheckAuthorityWindow(auth, 1500); err != nil {
		t.Errorf("mid-window must not error: %v", err)
	}
	if err := CheckAuthorityWindow(auth, 1000-60); err != nil {
		t.Errorf("inclusive lower skew edge must not error: %v", err)
	}
	for _, badNow := range []int64{1000 - 61, 2000 + 61} {
		err := CheckAuthorityWindow(auth, badNow)
		if err == nil {
			t.Errorf("expected AuthorityExpired at now=%d", badNow)
			continue
		}
		if _, ok := err.(*AuthorityExpired); !ok {
			t.Errorf("expected *AuthorityExpired, got %T", err)
		}
	}
}

func TestAuthorityEffectLabel(t *testing.T) {
	auth, err := GrantAuthorization(targetStateFixture, authorityStateFixture, 1000, 2000, "key_admin", "")
	if err != nil {
		t.Fatalf("GrantAuthorization: %v", err)
	}
	if got := AuthorityEffectLabel(nil, map[string]any{"effect_digest": "x"}); got != "" {
		t.Errorf("nil authorization -> want \"\", got %q", got)
	}
	if got := AuthorityEffectLabel(auth, map[string]any{}); got != UnverifiedAuthority {
		t.Errorf("no effect_digest -> want %q, got %q", UnverifiedAuthority, got)
	}
	matchDigest, _ := JSONDigest(targetStateFixture)
	if got := AuthorityEffectLabel(auth, map[string]any{"effect_digest": matchDigest}); got != AuthorizedMatch {
		t.Errorf("matching effect_digest -> want %q, got %q", AuthorizedMatch, got)
	}
	other := map[string]any{"ticket": "8842", "status": "refunded", "refund_cents": float64(7500)}
	mismatchDigest, _ := JSONDigest(other)
	if got := AuthorityEffectLabel(auth, map[string]any{"effect_digest": mismatchDigest}); got != AuthorizedMismatch {
		t.Errorf("differing effect_digest -> want %q, got %q", AuthorizedMismatch, got)
	}
}

// TestAuthorityReproducesFixedVector cross-checks against
// test-vectors.json -> authorization's own validity_window and effect_check
// fixed numbers — the same numbers the Python and TypeScript suites check —
// so all three SDKs are proven to agree on [TAP-AUTHORITY-VALIDITY] and
// [TAP-AUTHORITY-EFFECT], not merely that each independently satisfies its
// own logic.
func TestCheckAuthorityNotRevokedBoundary(t *testing.T) {
	if err := CheckAuthorityNotRevoked("auz_x", "sha256:v", nil, 1500); err != nil {
		t.Errorf("nothing revoked must not error: %v", err)
	}
	boundary := int64(2000)
	if err := CheckAuthorityNotRevoked("auz_x", "sha256:v", &boundary, 1999); err != nil {
		t.Errorf("before boundary must not error: %v", err)
	}
	for _, badTS := range []int64{2000, 2001} {
		err := CheckAuthorityNotRevoked("auz_x", "sha256:v", &boundary, badTS)
		if err == nil {
			t.Errorf("expected AuthorityRevoked at signedTS=%d", badTS)
			continue
		}
		revoked, ok := err.(*AuthorityRevoked)
		if !ok {
			t.Errorf("expected *AuthorityRevoked, got %T", err)
			continue
		}
		if revoked.RevokedID != "auz_x" || revoked.RevokedAt != 2000 || revoked.SignedTS != badTS {
			t.Errorf("unexpected fields: %+v", revoked)
		}
	}
}

func TestAuthorityReproducesFixedVector(t *testing.T) {
	repoRoot := findRepoRoot(t)
	data, err := os.ReadFile(filepath.Join(repoRoot, "test-vectors.json"))
	if err != nil {
		t.Fatalf("read test-vectors.json: %v", err)
	}
	var vectors map[string]any
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("parse test-vectors.json: %v", err)
	}

	av := vectors["authorization"].(map[string]any)
	signedEvent := av["signed_event"].(map[string]any)
	authRecord := signedEvent["authorization"].(map[string]any)
	auth, err := AuthorizationFromRecord(authRecord)
	if err != nil {
		t.Fatalf("AuthorizationFromRecord: %v", err)
	}

	vw := av["validity_window"].(map[string]any)
	acceptedNow, _ := asInt64(vw["accepted_now"])
	if err := CheckAuthorityWindow(auth, acceptedNow); err != nil {
		t.Errorf("vector's accepted_now must not error: %v", err)
	}
	for _, key := range []string{"rejected_now_not_yet_valid", "rejected_now_expired"} {
		badNow, _ := asInt64(vw[key])
		if err := CheckAuthorityWindow(auth, badNow); err == nil {
			t.Errorf("vector's %s=%d must raise AuthorityExpired", key, badNow)
		}
	}

	ec := av["effect_check"].(map[string]any)
	matchDigest := ec["authorized_match_effect_digest"].(string)
	mismatchDigest := ec["authorized_mismatch_effect_digest"].(string)
	if got := AuthorityEffectLabel(auth, map[string]any{"effect_digest": matchDigest}); got != AuthorizedMatch {
		t.Errorf("vector's authorized_match_effect_digest -> want %q, got %q", AuthorizedMatch, got)
	}
	if got := AuthorityEffectLabel(auth, map[string]any{"effect_digest": mismatchDigest}); got != AuthorizedMismatch {
		t.Errorf("vector's authorized_mismatch_effect_digest -> want %q, got %q", AuthorizedMismatch, got)
	}
	if got := AuthorityEffectLabel(auth, map[string]any{}); got != UnverifiedAuthority {
		t.Errorf("absent effect_digest -> want %q, got %q", UnverifiedAuthority, got)
	}
}
